Выполнены доработки по заданию спринта №27 (UPD-1)

Neolight79 opened the pull request:
- Исправлен баг с крашем при выходе из экрана плеера, если нет разрешений на показ уведомлений. Кроме того добавлен механизм открытия экрана настроек для разрешения уведомлений при нажатии на кнопку Play.
- На экране списка плейлистов, при возвращении из экрана создания плейлиста, список плейлистов обновляется, новый добавляется сверху.
- На экране избранного, при возвращении на него, список избранных треков обновляется, новый добавляется сверху.
- Вместо перебора индекса для повторяющихся имен файлов изображений обложки плейлиста генерирую GUID, что позволяет избавиться от повторного контроля существования файла.
- Добавлен контроль создания плейлиста с именем без значащих символов. Только из пробелов название плейлиста состоять не может.
- Оптимизированы алгоритмы на экране поиска треков. Текст запроса и список найденных треков сохраняется после возвращения из экрана плеера.
- Во всех местах, где требуется контроль окончаний для описания числовых значений перевел на использование plurals.
- Для всех экранов, где возможна ситуация закрытия самого экрана до закрытия модальных элементов добавлен механизм ожидания закрытия модальных элементов. Например на экране плейлиста.
- Поправил направление сдвига экранов при анимации для корректного сочетания направлений.

Выполнены задачи по заданию спринта №27
- Выполнена полная конверсия приложения на фреймворк Compose.
- Работа заняла 5 недель.
- Все View, фрагменты, стили, макеты layout, убраны из проекта и заменены на Compose функции. Также удалены неиспользуемые ресурсы.
- Объекты LiveData заменены на StateFlow.
- Навигация переведена на библиотеку Navigation.Compose.
- Слои data, domain и модули Koin практически не подверглись изменениям, в некоторых случаях изменились типы свойств моделей в связи с переходом на другие классы.
- Изменился подход к внедрению зависимостей для передачи ViewModel в Compose функции. Реализовано через koinViewModel() с передачей параметров через эту же функцию.
- Весь функционал и дизайн приложения сохранен.

Commit:
Выполнены задачи по заданию спринта №27
- Выполнена полная конверсия приложения на фреймворк Compose.
- Работа заняла 5 недель.
- Все View, фрагменты, стили, макеты layout, убраны из проекта и заменены на Compose функции. Также удалены неиспользуемые ресурсы.
- Объекты LiveData заменены на StateFlow.
- Навигация переведена на библиотеку Navigation.Compose.
- Слои data, domain и модули Koin практически не подверглись изменениям, в некоторых случаях изменились типы свойств моделей в связи с переходом на другие классы.
- Изменился подход к внедрению зависимостей для передачи ViewModel в Compose функции. Реализовано через koinViewModel() с передачей параметров через эту же функцию.
- Весь функционал и дизайн приложения сохранен.

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/main/ui/compose/MainScreen.kt:
```diff
@@ -0,0 +1,239 @@
+                    slideOutOfContainer(
+                        AnimatedContentTransitionScope.SlideDirection.Left, tween(300))
+                }) { backStackEntry ->
+                val trackID = backStackEntry.arguments?.getInt("trackID") ?: 0
```
Рекомендация:
лучше хранить ключи в константах

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/domain/models/DialogState.kt:
```diff
@@ -0,0 +1,10 @@
+package com.example.playlistmaker.media.domain.models
+
+sealed class DialogState {
```
здесь лучше подойдет sealed interface
https://stackoverflow.com/questions/69100358/what-is-the-difference-between-sealed-class-vs-sealed-interface-in-kotlin

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/compose/FavoriteScreen.kt:
```diff
@@ -0,0 +1,177 @@
+                    ),
+                    modifier = Modifier.weight(1F)
+                )
+                Icon(modifier = Modifier.width(13.dp),
```
Рекомендация:
В дизайне могут быть нечетные размеры, или не симметричные. Но лучше использовать в коде размеры, кратные 4 или 8. Документация - https://material.io/design/layout/spacing-methods.html#baseline-grid

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/compose/PlaylistsScreen.kt:
```diff
@@ -0,0 +1,141 @@
+            )
+        )
+        // Количество треков
+        val tracksQuantityString: String =
```
Рекомендация:
Воспользуйся лучше plurals, в андроиде уже реализована логика, с помощью которой для каждой локали будет подбираться соответствующее окончание. Почитать о том, как это использовать, можно в официальной документации - https://developer.android.com/guide/topics/resources/string-resource#Plurals

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/view_model/ManagePlaylistViewModel.kt:
```diff
@@ -0,0 +1,167 @@
+        val fileName = getFileNameFromUri(uri, context)
+        var file = File(filePath, "$fileName.jpg")
+        var i = 1
+        while (file.exists()) {
```
Рекомендация:
Тут можно сильно упростить, вместо простого индекса использовать текущее время в миллисекундах, тогда перебирать не придется

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/search/ui/compose/SearchScreen.kt:
```diff
@@ -0,0 +1,342 @@
+fun SearchField(viewModel: SearchViewModel, triggerDirectSearch: StateFlow<Long>) {
+
+    // Отслеживаем прилет состояния для запуска поиска
+    val triggerDirectSearchState = triggerDirectSearch.collectAsState().value
```
`collectAsStateWithLifecycle` - предпочтительнее, т.к. безопасно обрабатывает жизненный цикл

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/search/ui/compose/SearchScreen.kt:
```diff
@@ -0,0 +1,342 @@
+
+    // Триггер для выполнения принудительного поиска треков для команды от кнопки повтора попытки
+    val mTriggerDirectSearch = MutableStateFlow<Long>(0L)
+    val triggerDirectSearch: StateFlow<Long> = mTriggerDirectSearch.asStateFlow()
```
Рекомендация:
на самом деле для повторного запроса это будет этакий оверхед, лучше реализовать через одноразовый ивент

```
private val _searchRetry = MutableSharedFlow<Unit>(replay = 0)
val searchRetry = _searchRetry.asSharedFlow()

fun retrySearch() {
    viewModelScope.launch {
        _searchRetry.emit(Unit)
    }
}
```
и на стороне ui -> 
```
LaunchedEffect(Unit) {
    viewModel.searchRetry.collect {
        viewModel.search(viewModel.searchText.value) // повторный вызов
    }
}
```

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/search/ui/compose/SearchScreen.kt:
```diff
@@ -0,0 +1,342 @@
+        // Теперь обработчик стейта от ViewModel
+        when (searchState) {
+            // Если получаем объект Init, значит ничего показывать не надо
+            SearchState.Init -> { }
```
Рекомендация:
`SearchState.Init -> Unit`

sukhoikms27 reviewed (commented):
1) креш при отсутствии пермишена на нотификации:
```
FATAL EXCEPTION: main
Process: com.example.playlistmaker, PID: 24084
java.lang.IllegalArgumentException: Service not registered: com.example.playlistmaker.util.MusicServiceConnection@17a036f
	at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:2013)
	at android.app.ContextImpl.unbindService(ContextImpl.java:2197)
	at android.content.ContextWrapper.unbindService(ContextWrapper.java:925)
	at android.content.ContextWrapper.unbindService(ContextWrapper.java:925)
	at com.example.playlistmaker.player.ui.compose.PlayerScreenKt.unbindMusicService(PlayerScreen.kt:289)
	at com.example.playlistmaker.player.ui.compose.PlayerScreenKt$PlayerScreen$lambda$22$lambda$21$$inlined$onDispose$1.dispose(Effects.kt:68)
	at androidx.compose.runtime.DisposableEffectImpl.onForgotten(Effects.kt:87)
	at androidx.compose.runtime.internal.RememberEventDispatcher.dispatchRememberObservers(RememberEventDispatcher.kt:158)
	at androidx.compose.runtime.CompositionImpl.applyChangesInLocked(Composition.kt:1044)
	at androidx.compose.runtime.CompositionImpl.applyChanges(Composition.kt:1067)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:684)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:591)
	at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:39)
	at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.android.kt:108)
	at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.android.kt:41)
	at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1404)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1415)
	at android.view.Choreographer.doCallbacks(Choreographer.java:1015)
	at android.view.Choreographer.doFrame(Choreographer.java:941)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1389)
	at android.os.Handler.handleCallback(Handler.java:959)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8705)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [androidx.compose.runtime.PausableMonotonicFrameClock@683a068, androidx.compose.ui.platform.MotionDurationScaleImpl@95b2c81, StandaloneCoroutine{Cancelling}@277126, AndroidUiDispatcher@ea8f167]

```

sukhoikms27 commented:
> * Выполнена полная конверсия приложения на фреймворк Compose.
> * Работа заняла 5 недель.
> * Все View, фрагменты, стили, макеты layout, убраны из проекта и заменены на Compose функции. Также удалены неиспользуемые ресурсы.
> * Объекты LiveData заменены на StateFlow.
> * Навигация переведена на библиотеку Navigation.Compose.
> * Слои data, domain и модули Koin практически не подверглись изменениям, в некоторых случаях изменились типы свойств моделей в связи с переходом на другие классы.
> * Изменился подход к внедрению зависимостей для передачи ViewModel в Compose функции. Реализовано через koinViewModel() с передачей параметров через эту же функцию.
> * Весь функционал и дизайн приложения сохранен.

Работа супер!

Commit:
Выполнены доработки по заданию спринта №27 (UPD-1)
- Исправлен баг с крашем при выходе из экрана плеера, если нет разрешений на показ уведомлений. Кроме того добавлен механизм открытия экрана настроек для разрешения уведомлений при нажатии на кнопку Play.
- На экране списка плейлистов, при возвращении из экрана создания плейлиста, список плейлистов обновляется, новый добавляется сверху.
- На экране избранного, при возвращении на него, список избранных треков обновляется, новый добавляется сверху.
- Вместо перебора индекса для повторяющихся имен файлов изображений обложки плейлиста генерирую GUID, что позволяет избавиться от повторного контроля существования файла.
- Добавлен контроль создания плейлиста с именем без значащих символов. Только из пробелов название плейлиста состоять не может.
- Оптимизированы алгоритмы на экране поиска треков. Текст запроса и список найденных треков сохраняется после возвращения из экрана плеера.
- Во всех местах, где требуется контроль окончаний для описания числовых значений перевел на использование plurals.
- Для всех экранов, где возможна ситуация закрытия самого экрана до закрытия модальных элементов добавлен механизм ожидания закрытия модальных элементов. Например на экране плейлиста.
- Поправил направление сдвига экранов при анимации для корректного сочетания направлений.
Выполнены задачи по заданию спринта №27
- Выполнена полная конверсия приложения на фреймворк Compose.
- Работа заняла 5 недель.
- Все View, фрагменты, стили, макеты layout, убраны из проекта и заменены на Compose функции. Также удалены неиспользуемые ресурсы.
- Объекты LiveData заменены на StateFlow.
- Навигация переведена на библиотеку Navigation.Compose.
- Слои data, domain и модули Koin практически не подверглись изменениям, в некоторых случаях изменились типы свойств моделей в связи с переходом на другие классы.
- Изменился подход к внедрению зависимостей для передачи ViewModel в Compose функции. Реализовано через koinViewModel() с передачей параметров через эту же функцию.
- Весь функционал и дизайн приложения сохранен.
